internal/token/token_csp.go

```diff
@@ -0,0 +1,161 @@
+	"time"
+)
+
+type CSPAuthorizeResponse struct {
+	IdToken      string `json:"id_token"`
+	TokenType    string `json:"token_type"`
+	ExpiresIn    int    `json:"expires_in"`
+	Scope        string `json:"scope"`
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+}
+
+func (t *CspServerToServerTokenService) GetToken() string {
+
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	if !t.tokenReady {
+		cspAccessToken, err := t.CallCSP()
+
+		if err != nil {
+			return "INVALID_TOKEN"
+		}
+
+		t.AccessToken = cspAccessToken
+		t.tokenReady = true
+	}
+
+	return t.AccessToken
+}
+
+func (t *CspServerToServerTokenService) CallCSP() (string, error) {
+	var oauthPath = "/csp/gateway/am/api/auth/authorize"
+	client := &http.Client{}
+
+	requestBody := url.Values{"grant_type": {"client_credentials"}}.Encode()
+	req, err := http.NewRequest("POST", t.CSPBaseUrl+oauthPath, strings.NewReader(requestBody))
+
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Add("Authorization", encodeCSPCredentials(t.CSPClientId, t.CSPClientSecret))
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := client.Do(req)
+
+	if err != nil {
+		return "", err
+	}
+
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+
+	var cspResponse CSPAuthorizeResponse
+
+	err = json.Unmarshal(body, &cspResponse)
+
+	if !hasDirectIngestScope(cspResponse.Scope) {
+		log.Println("The CSP response did not find any scope matching 'aoa:directDataIngestion' which is required for Wavefront direct ingestion.")
+	}
+
+	if err != nil {
+		return "", err
+	}
+
+	t.startOrResetTicker(cspResponse)
+
+	log.Println("ACCESS TOKEN: " + cspResponse.AccessToken)
+	return cspResponse.AccessToken, nil
+}
+
+func hasDirectIngestScope(scope string) bool {
+	if len(scope) == 0 {
+		return false
+	}
+
+	for _, s := range strings.Split(scope, " ") {
+		if strings.Contains(s, "aoa:directDataIngestion") || strings.Contains(s, "aoa/*") || strings.Contains(s, "aoa:*") {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (t *CspServerToServerTokenService) startOrResetTicker(cspResponse CSPAuthorizeResponse) {
+	tickerDelay := t.getTickerDelay(cspResponse.ExpiresIn)
+
+	if t.ticker == nil {
+		t.ticker = time.NewTicker(time.Duration(tickerDelay) * time.Second)
+		t.done = make(chan bool)
+
+		// Goroutine
+		go func() {
+			log.Printf("Scheduling a goroutine to fetch fresh CSP credentials in %d seconds\n", tickerDelay)
+			for {
+				select {
+				case <-t.done:
+					return
+				case tick := <-t.ticker.C:
+					t.mutex.Lock()
+
+					log.Println("Re-fetching CSP credentials at", tick)
+					cspAccessToken, e := t.CallCSP()
+
+					if e != nil {
+						t.AccessToken = "INVALID_TOKEN"
+					} else {
+						t.AccessToken = cspAccessToken
+					}
+
+					t.mutex.Unlock()
+				}
+			}
+		}()
+	} else {
+		t.ticker.Reset(time.Duration(tickerDelay) * time.Second)
+	}
+}
+
+func (t *CspServerToServerTokenService) getTickerDelay(expiresIn int) int {
```

Comment:
change: method name

Reply:
I would expect a 'get' to take no arguments and return a state value. I want to look at where this is used and see what its intention is. Maybe 'normalize ticker delay?'